src/app/ClusterStateCache.h

```diff
@@ -310,12 +371,90 @@ class AttributeCache : protected ReadClient::Callback
         return CHIP_NO_ERROR;
     }
 
+    /*
+     * Execute an iterator function that is called for every event in the event data cache.
+     * This iterator is called in increasing order from the event with the lowest event number to the highest.
+     *
+     * The function is passed a const reference to the EventHeader associated with that event.
+     *
+     * The iterator is expected to have this signature:
+     *      CHIP_ERROR IteratorFunc(const EventHeader &eventHeader);
+     *
+     * Notable return values:
+     *      - If func returns an error, that will result in termination of any further iteration over events
+     *        and that error shall be returned back up to the original call to this function.
+     *
+     */
+    template <typename IteratorFunc>
+    CHIP_ERROR ForEachEventData(IteratorFunc func)
```

Comment:
So just thinking about API surface here, I would have expected APIs for:
1. Enumerate all events at a given path, possibly starting after some event number.
2. Init the cache with a given `mHighestReceivedEventNumber` (so we can reproduce the state we would be in after calling `ClearEventCache(false)` while creating a new cache).
3. Enumerate all events starting after a given event number.

I guess item 3 can be done with ForEachEventData and filtering in the consumer, so not that urgent.... Item 1 can also be done via filtering in the consumer, but feels like it would be a bit more work to do that filtering.